### ISSUE TITLE
"Any" tag `{...}:`

### ISSUE THREAD

**Issue by Y-Less** [Southclaws/pawn#264]
Would be useful for functions like `CallLocalFunction` where the varargs can be anything:

```pawn
native CallLocalFunction(const func[], const types[], {...}:...);
```

Don't know if `{...}:` is the best syntax, though it has the advantage of being similar to existing syntax for varargs, "rest" in array initialisation, and groups of tags; plus the advantage of not being a symbol like `Any;` which could be in use elsewhere.  The downsides being it is a little verbose and `{...}:...` is confusing.

I gave an example with varargs, but that isn't the only version.  Also `{...}:a, b = tagof (a)` should return the original tag, not a special representation of `{...}`.

**Comment by Southclaws** [comment]
I actually thought this was a feature in Pawn up until a few months ago when I finally tried to write a logger that took any tag. I thought the syntax was actually `{Float,...}:...` for some reason! This would be very useful for variadic funcs and I don't expect it to be super difficult to implement...

**Comment by Y-Less** [comment]
I don't know how special-cased it should be.  `{Float, ...}:` would be floats and everything (including floats), redundant but not wrong.  However, since in that case `{}` is a list of tags it would imply that the tag was just `...` not `{...}` implying this should also work: `Func(...:a, b)`.  I think it is better as strictly `{...}:` not just `...:`, so no others can be specified in the braces, and the dots as a tag can't be used without the braces.

Or maybe, since `_:` is already a special-case named tag, have `@:` as a similar special-case for "all".  That would more closely follow existing rules for symbols, and somewhat match the existing special tag.

**Comment by Southclaws** [comment]
Ah, `@` actually makes a lot more sense to me now compared to `...` which kind of feels like a continuation of what precedes it rather than just a catch-all expression.

**Comment by Y-Less** [comment]
The syntax is just random ideas, I'm open to others.

**Comment by YashasSamaga** [comment]
The compiler doesn't seem to allow `@:` tag. I have no idea why.

**Comment by IllidanS4** [comment]
I'd also like to see this implemented. I had an idea for the syntax of `{*}:`, as a sort of a wildcard. Maybe something to distringuish weak/strong tags would be interesting as well.

**Comment by Y-Less** [comment]
What would distinguishing weak/strong tags do?

**Comment by Y-Less** [comment]
I'm not convinced about `{*}` though - if we're reusing an existing operator, may as well stick to one that already means "more" (i.e. `...`), or something else entirely.  I'm actually now less enamoured by `@` for the same reason.  I also thought of `{}:`, but that's too easy to et by mistake.

**Comment by IllidanS4** [comment]
Probably nothing useful. I was thinking at first about some operation that would be more "valid" for weakly tagged values, but I realized that an untagged parameter would work the same.

Something the `{*}:` syntax may allow for would be an additional syntax of `{+}:` meaning "only tagged" (i.e. no `_:`). Not sure about the use cases again, just a random thought.

**Comment by Y-Less** [comment]
A feature should need a very good use case for introduction imho, not just because it looked neat.

**Comment by YashasSamaga** [comment]
1. what should this be called? wildcard tag?
2. what should the tag identifier for this wildcard tag be?

**Comment by Southclaws** [comment]
Wildcard sounds good, `...` looks fine, I'm not entirely sure how the tag value system works so I can't comment on that. A new warning would be kinda useful but not absolutely necessary.

**Comment by YashasSamaga** [comment]
The other alternatives to adding (adding new warnings/errors is bad imo) a new warning is to:
- `{Float, ...}` triggers an error saying unexpected `...`
- `{..., Float}` triggers an error saying `}` was expected

**Comment by Southclaws** [comment]
Yeah I think that makes more sense, I doubt many people are gonna do that anyway. (apart from [me...](https://github.com/pawn-lang/compiler/issues/264#issuecomment-358620265))

**Comment by IllidanS4** [comment]
I would design the system similarly to bounded vs. unbounded arrays and `sizeof`.
```pawn
native set_value({...}:val, tag_id = tagof(val));

stock set_value_ex({...}:val, tag_id = tagof(val))
{
    set_value(val); //should raise "indeterminate variable tag"
}
```
Does the wildcard tag have any use besides in parameters? If not, I find no use in `tagof(...)`.

**Comment by YashasSamaga** [comment]
```
f1({...}:arg) {
	#pragma unused arg
}


main() {
	f1(Tag:0);
	f1(0);
	f1(tag:0);
	
	const tagval = tagof(...);
	#pragma unused tagval
}

f2({..., Float}:arg) {

}

f3({Float, ...}:arg) {

}

#pragma unused f2
#pragma unused f3
```

```
(15) : error 001: expected token: "}", but found ","
(19) : error 001: expected token: "-tagname-", but found "wildcard tag"
```

**Comment by IllidanS4** [comment]
Well, multi-tagged variables are already supported only for parameters, so I think we should treat `{...}` as being equivalent to a set of all possible tags, like `{_,Float,...}`.

Which reminds me of this:
```pawn
native set_value({_,Float}:val, tag_id = tagof(val));

stock set_value_ex({_,Float}:val, tag_id = tagof(val))
{
    set_value(val);
}
```
This should basically raise the same warning, instead of silently passing the id of the first tag. Has this been addressed already?

**Comment by Y-Less** [comment]
Yes, this would only be for parameters, giving a variable any tag makes no sense. And it doesn't need an ID either, as @IllidanS4 said, it is literally just shorthand for `{every, tag, ever }:`, so should be treated as such - that can't be used for variable declarations and isn't given its own ID.

**Comment by YashasSamaga** [comment]
```
f1({...}:arg) {
	#pragma unused arg
}

g1(arg) {
	#pragma unused arg
}

g1_1({Float}:arg) {
	#pragma unused arg
}

f2({...}:arg = Tag:5) {
	#pragma unused arg
}

g2(arg = Tag:5) { // warning 213: tag mismatch: expected tag none ("_"), but found "Tag"
	#pragma unused arg
}

g2_1({Float}:arg = Tag:5) { // warning 213: tag mismatch: expected tag "Float", but found "Tag"
	#pragma unused arg
}

f3({...}:arg = tag:5) {
	#pragma unused arg
}

g3(arg = tag:5) {
	#pragma unused arg
}

g3_1({Float, tag}:arg = tag:5) { // warning 213: tag mismatch: expected tag "Float", but found "Tag"
	#pragma unused arg
}

f4(const {...}:arg[] = {Tag:1}){
	#pragma unused arg
}

g4(const arg[] = {Tag:1}) { // warning 213: tag mismatch: expected tag none ("_"), but found "Tag"
	#pragma unused arg
}

f5(const {...}:arg[][] = {{Tag:1}}) {
	#pragma unused arg
}

g5(const arg[][] = {{Tag:1}}) { // warning 213: tag mismatch: expected tag none ("_"), but found "Tag"
	#pragma unused arg
}

f6({...}:&arg = Tag:5) {
	#pragma unused arg
}

g6(&arg = Tag:5) { // warning 213: tag mismatch: expected tag none ("_"), but found "Tag"
	#pragma unused arg
}

f7({...}:...) {

}

g7(...) {

}


main() {
	f1(Tag:0);
	f1(0);
	f1(tag:0);
	g1(Tag:0); // warning 213: tag mismatch: expected tag none ("_"), but found "Tag"
	g1(0);
	g1(tag:0);
	g1_1(Tag:0);// warning 213: tag mismatch: expected tag "Float", but found "Tag"
	g1_1(Float:0);
	g1_1(0);// warning 213: tag mismatch: expected tag "Float", but found none ("_")
	g1_1(tag:0); // warning 213: tag mismatch: expected tag "Float", but found "tag"
	
	f2();
	f2(Tag:0);
	g2();
	g2(Tag:0); // warning 213: tag mismatch: expected tag none ("_"), but found "Tag"
	g2_1();
	g2_1(Tag:0); // warning 213: tag mismatch: expected tag "Float", but found "Tag"
	
	f3();
	f3(Tag:0);
	g3();
	g3(Tag:0); // warning 213: tag mismatch: expected tag none ("_"), but found "Tag"
	g3_1();
	g3_1(Tag:0); // warning 213: tag mismatch: expected tags "Float", or "tag"; but found "Tag"

	f4();
	f4({Tag:0});
	g4();
	g4({Tag:0});

	new Tag:tmp[][] = {{0}}; // warning 213: tag mismatch: expected tag "Tag", but found none ("_")
	f5();
	f5(tmp);
	g5();
	g5(tmp); // warning 213: tag mismatch: expected tag none ("_"), but found "Tag"
	
	f6();
	f6(tmp[0][0]);
	g6();
	g6(tmp[0][0]); // warning 213: tag mismatch: expected tag none ("_"), but found "Tag"
	
	f7(Tag:5);
	g7(Tag:5); // warning 213: tag mismatch: expected tag none ("_"), but found "Tag"
}

new arr1[5];
new arr2[] = {0};
new arr3[] = Tag:{ 0 };

#pragma unused arr1
#pragma unused arr2
#pragma unused arr3

f15({..., Float}:arg) { // error 001: expected token: "}", but found ","

}

f16({Float, ...}:arg) { // error 001: expected token: "-tagname-", but found "..."

}

#pragma unused f15
#pragma unused f16

enum ITEM {
	Float:item
};
new arr4[ITEM] = {Tag:0};
#pragma unused arr4
```

**Comment by Y-Less** [comment]
It occurred to me - Why not skip the braces? They are used for a set of tags, not a single one as this is, and there were already issues discussed about having other tags in the braces as well. Then I saw it and answered my own question:

```
...:...
```

**Comment by Southclaws** [comment]
That being said, the asterisk notation looks a little nicer! `*:...`.

We could also go with some mathematical notation... `∑:...` maybe even infinity to indicate infinite tags `∞:...` or maybe the set of possible tags `𝕌:...` or maybe one final math joke `ℵ₀:...`

**Comment by YashasSamaga** [comment]
`{...}:arg`

what should the compiler assume about the tag of `arg` for overload resolution?

The PR addressing this issue assumes `_:`. It does not allow `{Tag, ...}:`  but this syntax could be a way to specify what overloads should be used.

**Comment by Y-Less** [comment]
There are two options:

1. Just use `_`, since there is no way to do otherwise.

2. Don't allow operators at all, since it is mainly for natives and low-level operations anyway.

**Comment by IllidanS4** [comment]
What is the status of this? Are there any major issues that prevent this from being included in the next version? Any estimates?

**Comment by Y-Less** [comment]
The PR seems fairly complete.  We can always add overload resolution via `{Float, ...}:` later.

**Comment by YashasSamaga** [comment]
The issue with the PR is that it adds a new argument to few functions. You now there are two things to keep track of: wildcard bool and tag id.

I have got a new idea to use a variant for tags but this would involve a bit of work.

**Comment by Y-Less** [comment]
Why is that an issue?
